ansible/roles/vm_set/files/mux_simulator.py

```diff
@@ -116,370 +112,535 @@ def run_cmd(cmdline):
+        flows = {}
+        flows[self.ports[NIC]] = [
+            {'action': OUTPUT, 'out_port': self.ports[out_side]} for out_side in self.flows['upstream']['out_sides']
+        ]
+
+        if self.flows['downstream']['in_side'] is not None:
+            in_side = self.flows['downstream']['in_side']
+            in_port = self.ports[in_side]
+            flows[in_port] = [
+                {'action': OUTPUT, 'out_port': self.ports[out_side]} for out_side in self.flows['downstream']['out_sides']
+            ]
+
+        healthy = True
+        if len(self.flows['downstream']['out_sides']) != 1 or len(self.flows['upstream']['out_sides']) != 2:
+            healthy = False
+
+        status = {
+            'bridge': self.bridge,
+            'vm_set': self.vm_set,
+            'port_index': self.port_index,
+            'ports': self.ports,
+            'active_port': self.active_port,
+            'active_side': self.active_side,
+            'standby_side': self.standby_side,
+            'standby_port': self.standby_port,
+            'flows': flows,
+            'flap_counter': self.flap_counter,
+            'healthy': healthy
+        }
+        self.lock.release()
+        return status
+
+    def set_active_side(self, new_active_side):
+        """Set the active side of the mux bridge to the specified side.
+
+        If the specified side is same as the current active side of bridge, no config change is reuqired. Otherwise,
+        this method will run ovs-ofctl command to remove flow and add a new flow to switch active side. All the
+        related instance attributes are updated after open flow rules are changed.
+        """
+        try:
+            self.lock.acquire()
+
+            self.info('>>>>>> updating mux active side from {} to {}'.format(self.active_side, new_active_side))
+            if new_active_side == RANDOM:
+                new_active_side = random.choice([UPPER_TOR, LOWER_TOR])
+
+            if self.active_side == new_active_side:
+                self.info('current active_side={}, new_active_side={}, no need to change. <<<<<<'.format(self.active_side, new_active_side))
+                return
+
+            # Need to toggle active side
+            if new_active_side == TOGGLE:
+                new_active_side = UPPER_TOR if self.active_side == LOWER_TOR else LOWER_TOR
+
+            new_active_port = self.ports[new_active_side]
+
+            if len(self.flows['downstream']['out_sides']) == 1:
+                action_desc = '{}:"{}"'.format(OUTPUT, self.ports[NIC])
+                run_cmd(DEL_FLOW_CMD.format(
+                    self.bridge,
+                    self.active_port))
+                # Immediately update state after flow config changed to ensure consistency
+                self._init_active_standby(None)
+                self.flows['downstream']['in_side'] = self.active_side
+                self.flows['downstream']['out_sides'] = []
```

Comment:
What's the purpose for Line 376 - 383? Why do we need to remove NIC and then readd?

Reply:
To switch active side, the flow from ToR interface to PTF interface (NIC) need to be changed. To be more specific, the `in_port` of the flow needs to be changed to another ToR interface. To make such flow change, the only supported way is to remove it and add a new one. Command like `ovs-ofctl mod-flow` can only modify the `action` part of the flow, not the `in_port` part.

Reply:
I see. Thanks